### PR TITLE
[RLlib] added preprocessor to dataset

### PR DESCRIPTION
**Pull request by kouroshHakha**
Signed-off-by: Kourosh Hakhamaneshi <kourosh@anyscale.com>

## Why are these changes needed?

Dataset reader does not apply rllib preprocessors if the observation space is complex.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/offline/dataset_reader.py [hunk 1]
```diff
@@ -251,12 +252,34 @@ def next(self) -> SampleBatchType:
...
+                f"{len(preprocessors)} policy ids in the preprocessor dict."
+            )
+            assert len(preprocessors) == 1 and "default_policy" in preprocessors, msg
+            preprocessor = preprocessors["default_policy"]
+            for key in (SampleBatch.CUR_OBS, SampleBatch.NEXT_OBS):
+                if key in batch:
+                    samples = []
+                    for sample in batch[key]:
+                        samples.append(preprocessor.transform(sample))
+                    batch[key] = np.stack(samples, axis=0)
```
Comment (gjoliver) [inline rllib/offline/dataset_reader.py:280]:
```
batch[key] = np.stack([preprocessor.transform(s) for s in batch[key]])
```

File: rllib/offline/dataset_reader.py [hunk 2]
```diff
@@ -251,12 +252,34 @@ def next(self) -> SampleBatchType:
...
-            d = postprocess_actions(d, self._ioctx)
+            d = postprocess_actions(d, self._ioctx.worker)
             count += d.count
+            d = self._preprocess_if_needed(d)
             ret.append(self._postprocess_if_needed(d))
         ret = concat_samples(ret)
         return ret
 
+    def _preprocess_if_needed(self, batch: SampleBatchType) -> SampleBatchType:
+        # TODO: @kourosh, preprocessor is only supported for single agent case.
```
Comment (gjoliver) [inline rllib/offline/dataset_reader.py:263]: seems like I can replace all of these with connectors after it's enabled by default.

File: rllib/offline/dataset_reader.py [hunk 3]
```diff
@@ -251,12 +252,34 @@ def next(self) -> SampleBatchType:
             d = next(self._iter).as_pydict()
             # Columns like obs are compressed when written by DatasetWriter.
             d = from_json_data(d, self._ioctx.worker)
-            d = postprocess_actions(d, self._ioctx)
+            d = postprocess_actions(d, self._ioctx.worker)
             count += d.count
+            d = self._preprocess_if_needed(d)
```
Comment (gjoliver) [inline rllib/offline/dataset_reader.py:257]: should we re-arrange these to be:
```
...
count += d.count
d = self._preprocess_if_needed(d)
d = postprocess_actions(d, self._ioctx.worker)
d = self._postprocess_if_needed(d)
ret.append(d)
```
so it looks like preprocessing happens before postprocessing??

**Review by gjoliver — COMMENTED** [review]
a couple of nits.

File: rllib/offline/dataset_reader.py (hunk 3 above)
Comment (kouroshHakha) [reply rllib/offline/dataset_reader.py:262]: good idea.

File: rllib/offline/dataset_reader.py (hunk 2 above)
Comment (kouroshHakha) [reply rllib/offline/dataset_reader.py:270]: sg, this was not supported properly yet. So let's continue with this solution for now.

File: rllib/offline/dataset_reader.py (hunk 1 above)
Comment (kouroshHakha) [reply rllib/offline/dataset_reader.py]: sg.

**Comment by kouroshHakha** [comment]
@gjoliver I addressed all your comments.

**Review by gjoliver — APPROVED** [review]
looks nice. thanks.

**Pull request merged**